### PR TITLE
Support advanced indexing for chainer.Variable.__getitem__

### PR DESCRIPTION
**Pull request by yuyu2172**
Related #2202 #2201, #2175, #2134, #2126, #2099, #2043, #1863, #1832.
Resolves issue #2042. 

This PR supports advanced indexing for `chainer.Variable.__getitem__`.

Here are lists of functionalities that are added.
+ support for integer arrays. This includes multiple integer arrays (e.g. `a[i1, i2]`) and combination of integer arrays and basic indexing (e.g. `a[i1, 2:10, i2]`).
+ support for boolean array indexing.

This PR does not support multiple boolean arrays as indexing.

### CODE REVIEW

File: chainer/functions/array/get_item.py [hunk 1]
```diff
@@ -15,55 +13,66 @@ class GetItem(function.Function):
...
 
 def get_item(x, slices):
     """Extract elements from array with specified shape, axes and offsets.
 
     Args:
-        x (tuple of Variables): Variable to be sliced.
-        slices (int, slice, None or Ellipsis or tuple of them): Basic slicing
-            to slice a variable. It supports ``int``, ``slice``, ``newaxis``
-            (equivalent to ``None``) and ``Ellipsis``.
+        x (tuple of Variables): A variable to be sliced.
```
Comment (muupan) [inline chainer/functions/array/get_item.py]: Since this function accepts only one Variable, `(tuple of Variables)` should be `(~chainer.Variable)`, right?

File: chainer/functions/array/get_item.py [hunk 2]
```diff
@@ -15,55 +13,66 @@ class GetItem(function.Function):
...
+        slices (array_like or tuple): It is integer, slices, ellipsis,
+            numpy.newaxis, integer array-like, boolean array-like or tuple of
+            them.
```
Comment (muupan) [inline chainer/functions/array/get_item.py]: `(array_like or tuple)` doesn't include int, slice, None or Ellipsis. Maybe `(int, slice, None, Ellipsis or array_like or tuple of them)`?

**Comment by yuyu2172** [comment]
Thanks. I fixed the documentation.

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py [hunk 3]
```diff
@@ -76,6 +76,111 @@ def test_backward_gpu(self):
...
+        {'slices': ([1, 0], slice(1, 2), [0, 0]), 'sliced_shape': (2, 1)},
+        {'slices': ([[1, 1], [1, 0]], slice(1, 2), 1),
+         'sliced_shape': (2, 2, 1)},
+        {'slices': numpy.array([True] * 18 + [False] * 6).reshape(4, 3, 2),
+         'sliced_shape': (18,)},
+        {'slices': numpy.array([True, False, False, True]),
+         'sliced_shape': (2, 3, 2)},
+        {'slices': (slice(None), numpy.array([True, False, True])),
+         'sliced_shape': (4, 2, 2)},
+    ],
```
Comment (muupan) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: Could you add more corner cases such as empty list `[]` and all-False list?

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py [hunk 4]
```diff
@@ -76,6 +76,111 @@ def test_backward_gpu(self):
...
+
+class TestCupyIndicesGetItem(unittest.TestCase):
+
+    def setUp(self):
+        self.x_data = numpy.random.uniform(
+            -1, 1, (4, 3, 2)).astype(numpy.float32)
+        self.sliced_shape = (2, 2, 3)
+        self.gy_data = numpy.random.uniform(
+            -1, 1, self.sliced_shape).astype(numpy.float32)
+        self.slices = ([1, 0], slice(None), [[1, 1], [1, 1]])
```
Comment (muupan) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: `self.slices` must contain numpy.ndarray to test cupy.ndarray indices, right? I think it's better to test multiple advanced-indexing patterns here as well.

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py [hunk 5]
```diff
@@ -76,6 +76,118 @@ def test_backward_gpu(self):
...
+
+
+@parameterize(
+    {'slices': ([1, 0], [1, 1]), 'sliced_shape': (2, 2)},
+    {'slices': ([1, 0], slice(None), [[1, 1], [1, 1]]),
+     'sliced_shape': (2, 2, 3)},
+    {'slices': ([1, 0], [1, 1], [0, 0]), 'sliced_shape': (2,)},
+    {'slices': (slice(None), numpy.array([True, False, True])),
+     'sliced_shape': (4, 2, 2)},
+)
```
Comment (muupan) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: My understanding is:
- `TestCupyIndicesGetItem` is to test advanced indexing when `slices` contains `cupy.ndarray`.
- Since you cannot set `cupy.ndarray` directly inside `@parameterize`, you set `numpy.ndarray` and convert it to `cupy.ndarray` inside `@attr.gpu`.

Is that right? I still don't understand why here you also include test cases not containing `numpy.ndarray`, which seem to be already tested in `TestGetItemAdvanced`.

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py [hunk 6]
```diff
@@ -76,6 +76,118 @@ def test_backward_gpu(self):
...
+    ],
+    [
+        {'dtype': numpy.float32},
+        {'dtype': numpy.int32},
+        {'dtype': numpy.uint32},
+        {'dtype': numpy.uint64},
+        {'dtype': numpy.ulonglong},
+    ]
+))
+class TestGetitemAdvanced(unittest.TestCase):
```
Comment (muupan) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: Getitem -> GetItem?

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py (hunk 3 above)
Comment (yuyu2172) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: Thanks for pointing this out.
I found that `cupy.scatter_add` can not handle that kind of input. Could you review this PR after that fix is merged?

File: tests/chainer_tests/functions_tests/array_tests/test_get_item.py (hunk 5 above)
Comment (yuyu2172) [inline tests/chainer_tests/functions_tests/array_tests/test_get_item.py]: Your interpretation is correct. Thanks.

As you pointed out, the test cases with list (e.g. `[1,0]`) does not make sense in the current code.
However, this is due to my mistake in the test code. These lists are suppose to be converted into cupy arrays. I commited a change that does this, so could you please check it?

**Comment by yuyu2172** [comment]
Fix is still WIP.

**Comment by muupan** [comment]
@yuyu2172 Please let me know if it's ready

**Comment by yuyu2172** [comment]
@muupan 

I fixed the code. Could you check it?

File: chainer/functions/array/get_item.py [hunk 7]
```diff
@@ -15,55 +13,72 @@ class GetItem(function.Function):
...
 def get_item(x, slices):
     """Extract elements from array with specified shape, axes and offsets.
 
     Args:
-        x (tuple of Variables): Variable to be sliced.
-        slices (int, slice, None or Ellipsis or tuple of them): Basic slicing
-            to slice a variable. It supports ``int``, ``slice``, ``newaxis``
-            (equivalent to ``None``) and ``Ellipsis``.
+        x (~chainer.Variable): A variable to be sliced.
+        slices (int, slice, Ellipsis, None, integer array-like or tuple of\
```
Comment (muupan) [inline chainer/functions/array/get_item.py]: boolean array-like is missing

File: chainer/functions/array/get_item.py [hunk 8]
```diff
@@ -15,55 +13,72 @@ class GetItem(function.Function):
...
+        them):
+            It is integer, slices, ellipsis,
```
Comment (muupan) [inline chainer/functions/array/get_item.py]: s/slices/slice/ ?

**Comment by yuyu2172** [comment]
Thanks.

I fixed what you pointed out as well as some grammatical issues.

**Comment by muupan** [comment]
Thanks for your great work!

**Comment by yuyu2172** [comment]
Thanks for reviews!
I think we can close related issues (#1433, #1555).

**Comment by muupan** [comment]
Thanks for letting me know. I've closed those issues.
